pkg/controllers/nodeclass/capacityreservation.go

```diff
@@ -0,0 +1,133 @@
+	clk      clock.Clock
+	cm       *pretty.ChangeMonitor
+}
+
+func NewCapacityReservationReconciler(clk clock.Clock, provider capacityreservation.Provider) *CapacityReservation {
+	return &CapacityReservation{
+		provider: provider,
+		clk:      clk,
+		cm:       pretty.NewChangeMonitor(),
+	}
+}
+
+func (c *CapacityReservation) Reconcile(ctx context.Context, nc *v1.EC2NodeClass) (reconcile.Result, error) {
+	reservations, err := c.provider.List(ctx, nc.Spec.CapacityReservationSelectorTerms...)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting capacity reservations, %w", err)
+	}
+	if len(reservations) == 0 {
+		nc.StatusConditions().SetTrue(v1.ConditionTypeCapacityReservationsReady)
+		return reconcile.Result{RequeueAfter: capacityReservationPollPeriod}, nil
+	}
+
+	if ids := lo.Map(reservations, func(r *ec2types.CapacityReservation, _ int) string {
+		return *r.CapacityReservationId
+	}); c.cm.HasChanged(nc.Name, ids) {
+		log.FromContext(ctx).V(1).WithValues("ids", ids).Info("discovered capacity reservations")
+	}
+	sort.Slice(reservations, func(i, j int) bool {
+		return *reservations[i].CapacityReservationId < *reservations[j].CapacityReservationId
+	})
+	errors := []error{}
+	nc.Status.CapacityReservations = []v1.CapacityReservation{}
+	for _, r := range reservations {
+		reservation, err := capacityReservationFromEC2(r)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		nc.Status.CapacityReservations = append(nc.Status.CapacityReservations, reservation)
+	}
+	if len(errors) != 0 {
+		log.FromContext(ctx).Error(multierr.Combine(errors...), "failed to update status with %d of %d capacity reservations", len(errors), len(reservations))
+	}
+	nc.StatusConditions().SetTrue(v1.ConditionTypeCapacityReservationsReady)
+	return reconcile.Result{RequeueAfter: c.requeueAfter(reservations...)}, nil
+}
+
+func capacityReservationFromEC2(cr *ec2types.CapacityReservation) (v1.CapacityReservation, error) {
+	if !lo.Contains([]ec2types.InstanceMatchCriteria{
+		ec2types.InstanceMatchCriteriaOpen,
+		ec2types.InstanceMatchCriteriaTargeted,
+	}, cr.InstanceMatchCriteria) {
+		return v1.CapacityReservation{}, fmt.Errorf("capacity reservation %s has an unsupported instance match criteria %q", *cr.CapacityReservationId, cr.InstanceMatchCriteria)
+	}
+	var endTime *metav1.Time
+	if cr.EndDate != nil {
+		endTime = lo.ToPtr(metav1.NewTime(*cr.EndDate))
+	}
+
+	return v1.CapacityReservation{
+		AvailabilityZone:       *cr.AvailabilityZone,
+		AvailableInstanceCount: int(*cr.AvailableInstanceCount),
+		EndTime:                endTime,
+		ID:                     *cr.CapacityReservationId,
+		InstanceMatchCriteria:  string(cr.InstanceMatchCriteria),
+		InstanceType:           *cr.InstanceType,
+		OwnerID:                *cr.OwnerId,
+		TotalInstanceCount:     int(*cr.TotalInstanceCount),
+	}, nil
+}
+
+func (c *CapacityReservation) requeueAfter(reservations ...*ec2types.CapacityReservation) time.Duration {
```

Comment:
Can you throw a comment over this as well that describes the logic that is being handled here at a high-level -- mainly just about how we are dealing with the endTime and requeueing based on that